### PR TITLE
[vendor-change] Adding support for exporting logs from Stackdriver using Sink

### PR DESCRIPTION
**Pull request by gargnupur**

### CODE REVIEW

**Comment by ldemailly** [comment]
sgtm vendor wise - merge https://github.com/istio/vendor-istio/pull/54 once this is approved and green tests

File: mixer/doc/adapter-development-walkthrough.md [hunk 1]
```diff
@@ -488,10 +475,12 @@ spec:
 
 # Step 7: Start Mixer and validate the adapter.
 
-Start the mixer pointing it to the sample operator configuration
+Start the mixer pointing it to the testdata operator configuration
 
 ```bash
-pushd $MIXER_REPO && go install ./... && mixs server --configStoreURL=fs://$MIXER_REPO/adapter/mysampleadapter/sampleoperatorconfig
+pushd $ISTIO/istio && make mixs
+// locate mixs binary, should be $GOPATH/out/linux_amd64/release/mixs
```
Comment (douglas-reid) [inline mixer/doc/adapter-development-walkthrough.md]: nit: we should clarify that this is the output directory for certain linux architectures. we probably want a disclaimer for mac too.

File: mixer/doc/adapter-development-walkthrough.md [hunk 2]
```diff
@@ -423,26 +425,11 @@ Now your adapter is plugged into Mixer and ready to receive data.
...
-
-```bash
-cp $MIXER_REPO/testdata/config/attributes.yaml sampleoperatorconfig
-```
-
-
-Create a sample operator config file with name `config.yaml` inside the `sampleoperatorconfig` directory with the following content:
-
-Add the following content to the file `sampleoperatorconfig/config.yaml.`
+Add the following content to the file `$MIXER_REPO/testdata/config/mysampleadapter.yaml.`
```
Comment (douglas-reid) [inline mixer/doc/adapter-development-walkthrough.md]: my only concern here is that we will liter `testdata/config` with the adapter configs for all adapters, making it harder to opt out of loading handlers for local testing.

@guptasu thoughts here?

File: mixer/doc/adapter-development-walkthrough.md [hunk 3]
```diff
@@ -37,12 +37,13 @@ git clone https://github.com/istio/istio
 
 Install protoc (version 3.5.1 or higher) from [https://github.com/google/protobuf/releases](https://github.com/google/protobuf/releases) and add it to your PATH
 
-Set the MIXER_REPO variable to the path where the mixer repository is on the local machine. Example `export MIXER_REPO=$GOPATH/src/istio.io/istio/mixer`
+Set the MIXER_REPO variable to the path where the mixer repository is on the local machine. Example `export MIXER_REPO=$GOPATH/src/istio.io/istio/mixer`.
+Also, $ISTIO should point to $GOPATH/src/istio.io.
 
-Successfully build the repo.
+Successfully build the mixer repo.
```
Comment (douglas-reid) [inline mixer/doc/adapter-development-walkthrough.md]: Let's make this: `Successfully build the mixer server.`.

File: mixer/adapter/stackdriver/log/log.go [hunk 4]
```diff
@@ -106,14 +118,36 @@ func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler,
...
+			})
+			if err != nil {
+				if strings.Contains(err.Error(), "AlreadyExists") {
+					_, err = syncClient.UpdateSink(context.Background(), &logadmin.Sink{
+						ID:          log.SinkInfo.Id,
+						Destination: log.SinkInfo.Destination,
+						Filter:      log.SinkInfo.Filter,
+					})
+				}
+				if err != nil {
```
Comment (douglas-reid) [inline mixer/adapter/stackdriver/log/log.go]: nit: should this check be rolled inside the `if` above it (as that is the only way a subsequent check for `nil` is needed)?

maybe we could make this block clearer as follows:

```
if log.SinkInfo != nil {
  sink := &logadmin.Sink{...}
  ctx := context.Background()
  var sinkErr error
  if _, sinkErr = syncClient.CreateSink(ctx, sink); isAlreadyExists(sinkErr) {
     _, sinkErr = syncClient.UpdateSink(ctx, sink)
  }
  if sinkErr != nil {
     logger.Warningf(...)
  }
}

...

func isAlreadyExists(err error) bool {
   if err == nil {
      return false
   }
   return strings.Contains(err.Error(), "AlreadyExists")
}

File: mixer/doc/adapter-development-walkthrough.md (hunk 3 above)
Comment (gargnupur) [reply mixer/doc/adapter-development-walkthrough.md]: Thanks..Fixed

File: mixer/doc/adapter-development-walkthrough.md (hunk 1 above)
Comment (gargnupur) [reply mixer/doc/adapter-development-walkthrough.md]: Thanks.. Fixed

File: mixer/adapter/stackdriver/log/log.go (hunk 4 above)
Comment (gargnupur) [reply mixer/adapter/stackdriver/log/log.go]: Thanks.. yes that looks much better.

**Comment by ldemailly** [comment]
hey guys, you should not lgtm vendor changes

**Comment by gargnupur** [comment]
@ldemailly : should this be reverted now? I see a merge conflict in https://github.com/istio/vendor-istio/pull/54

**Comment by ldemailly** [comment]
let's try to resolve it in your PR and fix it forward, it's a bit hard to revert this one as more commits went in since

**Comment by douglas-reid** [comment]
shouldn't there be a presubmit explicitly preventing the bot from merging changes that have vendor changes? having 6 different ways of approving PRs is unmanageable for simpletons like me.

**Comment by ldemailly** [comment]
yes I need to find a way to tell the bot or a blocking build to not allow merges

**Pull request merged**